Echo

patrikkocsis opened the pull request.

patrikkocsis committed:
Echo done

patrikkocsis committed:
Makefile

oliverosz commented on directory.cpp:
```diff
@@ -1,36 +1,42 @@
 }
 
-void directory::ls()
+list<ancestor *> directory::getSubdirectories() const
```
Miért kell lemásolni ezt a listát? Jobb lenne referencia szerint átadni.

oliverosz commented on filesystem.cpp:
```diff
@@ -1,64 +1,99 @@
+    for(auto i:currentdir->getSubdirectories()){
+            if(i->getname()==fname){
+                file* tempfile=dynamic_cast<file*>(i);
+                tempfile->setContent(content);
```
Az egy dolog, hogy a `getSubdirectories()` adja vissza a fájlok listáját is, de ez így megengedi azt is, hogy egy könyvtárba echozzunk szöveget, és akkor elcrashel.

oliverosz reviewed (changes requested):
A minél kisebb commitok elvét nem sikerült tartani. Ebben az "Echo done"-nak nagyon kis része az echo megvalósítása.

A Makefile nem jó, az object fájlok dependency listájában az include-olt header fájlokat is fel kell sorolni.

A `cd ..` és `rm -rf` esetén a `..` és `-rf` az első argumentum, nem a parancs része, direkt van szóközzel elválasztva.

lebbenszki committed:
echo fixed

rf és .. első argumentumba került, makefile javítva, getSubdirectories() miatt nem crashel már + referencia szerint átadás

lebbenszki committed:
-const

lebbenszki committed:
config update

oliverosz commented on directory.cpp:
```diff
@@ -15,7 +15,7 @@ directory::~directory()
 }
 
-list<ancestor *> &directory::getSubdirectories() const
+list<ancestor *> &directory::getSubdirectories()
```
De miért nem `const`ez a függvény? Inkább konstans referenciát kéne visszaadni, és akkor engedi.

oliverosz commented on filesystem.cpp:
```diff
@@ -86,10 +88,17 @@ int filesystem::echo(string content,string fname){
-                return 1;
+                if(tempfile!=nullptr){
+                    tempfile->setContent(content);
+                    i=tempfile;
```
Erre nincs szükség, ha referencia szerint megy végig az elemeken: `auto &i`. Meg gyorsabb is, mert nem másolja le az elemeket feleslegesen.

oliverosz commented on filesystem.cpp:
```diff
@@ -137,15 +144,17 @@ void filesystem::start(){
-               else {
-                   currentdir->rm(argument1);
+               else if(argument1=="-rf"){
+                   cout<<argument2;
```
Ez a kiírás itt elég megtévesztő. Pl.:
````
cd a
ls
root/a/
   b
   c
cd ..
rm -rf a
ab torolve
c torolve
a torolve
````

oliverosz commented on directory.cpp:
```diff
@@ -61,60 +61,89 @@ void directory::echo(string content, string name){
+            //ha üres a mappa kitöröljük és továbbiterálunk
+                if(!temp->hasDirs()){
+                    subdirectories.erase(i++);
+                    return 1;
```
Nem iterálunk tovább, hanem visszatérünk a függvényből. Ahogy lentebb egy másik ágban is felesleges a léptetés, meg hibás a komment.

oliverosz commented on directory.cpp:
```diff
@@ -61,60 +61,89 @@ void directory::echo(string content, string name){
+                //ha nem üres meghívjuk a segedrmrf-et ami rekurzívan bejárja az adott mappát
+                else{
+                    temp->segedrmrf();
+                }
```
Itt sem kell továbbiterálni, mert ugyanazon névvel nem lehet másik fájl/mappa.

És miért is kell ez a rekurzív segedrmrf? Miért nem elég a szülő mappát kitörölni? (Ez most nem költői kérdés, várom rá a választ @Teaching-projects/gatewaytimeouterror )

szerdahelyir replied:
@oliverosz amikor nem rekurzívan próbáltuk törölni a szülőmappát, akkor a program lefagyott, amikor pedig rekurzívan a fastruktúra végpontjaival kezdtük a törlést, akkor minden működött rendesen(elvileg)

oliverosz replied:
@szerdahelyir A szülő törlését én úgy értem, hogy `subdirectories.erase(i)`, mint ahogy üres mappánál van fentebb, tehát attól nem kéne lefagynia. Ha nem üres a mappa, akkor miért nem lehet ugyanúgy kezelni?

szerdahelyir replied:
@oliverosz valóban működik így is 😄  de így nem lesz memóriaszivárgás, hogy abban a mappában voltak még fájlok/alkönyvtárak?

oliverosz replied:
@szerdahelyir Kb. ugyanannyi memóriaszivárgás lesz benne, mint a mostaniban. :D A segedrmrf-ben sem látok semmi delete hívást, szóval az ezen nem segít. Csak a listákból törli az elemeket, de azt a lista destruktora automatikusan is megcsinálná, ha a fájlok és könyvtárak destruktálódnának.

szerdahelyir replied:
@oliverosz így már világos :DD köszönjük a felvilágosítást, javítjuk a hibákat

oliverosz reviewed (changes requested):
Látható, hogy az rm, rmrf és segedrmrf függvények nagyon hasonló dolgot csinálnak. A kódduplikációt meg nem szeretjük. Jó lenne ezeket egy függvénybe egyesíteni, aminek egy bool paraméter mondaná meg, hogy a nem üres mappákat is törölje-e ki.

A hibakezelés sem teljesen jó, pl. az rm hibakóddal tér vissza, de az sehol sincs használva. Így most nem ír ki hibát, ha nem létező mappát próbálok törölni.

szerdahelyir committed:
rmrf fix error msg added

szerdahelyir committed:
file destructor fix

szerdahelyir committed:
deleting destructors

oliverosz reviewed (approved):
A conflictokat még ki kell javítani, aztán mergelhető.

szerdahelyir committed:
Merge branch 'master' into echo

szerdahelyir committed:
fixing mergecommit

szerdahelyir commented:
@oliverosz elvileg most már minden conflict javítva lett

oliverosz commented:
Ok, @hegyhati tudja bemergelni.
